【依頼】商品情報編集機能実装のコードレビュー

kokoro-kusakabe opened the pull request:
#フリマアプリにおける商品情報編集機能の実装をしました。
　コードレビューをお願いします。

#出品者Aさんが商品First,商品Secondを出品し商品Secondだけ購入した状態
を用意しました。

Aさんが商品firstを編集画面に遷移し元々合った情報が入力された状態で表示されているか
https://gyazo.com/f5c7ade56a6a2fb2e5819fc6be5086d1
Aさんが商品firstを編集画面に遷移しimage以外空にした時のエラーメッセージの表示
https://gyazo.com/3d75774da46d874dcbb38af0c696d8d5
Aさんが商品firstを編集しようとした時
https://gyazo.com/5e38fd4a80e4d913d227e4c7fb6b7b9e
Aさんが商品Secondを編集しようとした時
https://gyazo.com/f83bdeb13a912dcb6589839f77014437
Bさんが商品firstを編集しようとした時
https://gyazo.com/ad480a41ca3ce3675789d82983290a03
Bさんが商品Secondを編集しようとした時
https://gyazo.com/8e646c63682a9b2c6231d7ca06c919bb
ログアウトしてるユーザーが商品Firstを編集しようとした時
https://gyazo.com/14236f66a317be3374f7748aae483e57
ログアウトしてるユーザーが商品Secondを編集しようとした時
https://gyazo.com/755f833ffcf3f363f17ae2e26aa88ca8

kokoro-kusakabe committed:
ルーティングとコントローラーの設定

kokoro-kusakabe committed:
ビューの設定

kokoro-kusakabe committed:
エラーメッセージの表示

kokoro-kusakabe committed:
商品,ユーザーの状態でeditアクションをした時の動作

商品が売り切れ状態はトップページ遷移
本人以外に編集させない
本人の出品商品が売り切れていたら
トップページに遷移
ログアウトユーザーは編集画面に
遷移せずトップページに遷移

i-date commented on app/controllers/items_controller.rb:
```diff
@@ -20,6 +21,20 @@ def create
   end
 
+  def edit
+    @item = Item.find(params[:id])
```
25行目ですが、他アクションでも共通のコード`@item = Item.find(params[:id])`が見られます。
そのため、アクションに共通コードを記載するのではなく、DRY（Don't Repeat Youself）の原則（同じコードを同ファイルで繰り返し書かない）に従い一つのメソッドにまとめ、そのメソッドを必要なアクションの時にbefore_actionで呼び出しましょう。

また、before_actionでメソッドを呼び出していただくことで、メソッドにしたコードはアクション内には不要になりますので削除しましょう。

i-date commented on app/controllers/items_controller.rb:
```diff
@@ -20,6 +21,20 @@ def create
+  end
+
+  def update
+      @item = Item.find(params[:id])
```
29行目も、アクションに共通コードを記載するのではなく、一つのメソッドにまとめ、そのメソッドを必要なアクションの時にbefore_actionで呼び出しましょう。

i-date commented on app/controllers/items_controller.rb:
```diff
@@ -20,6 +21,20 @@ def create
+  def update
+      @item = Item.find(params[:id])
+    if  @item.update(item_params)
+        redirect_to "/items/#{@item.id}"
```
31行目ですが、リダイレクト先のpathにprefix（`～_path(引数)`、`～`と`引数`は適宜書き換え）を使用してはいかがでしょうか？
prefixを使用すると、任意のpathについて一意な命名になります。
それにより、pathがどこに遷移するのか分かりやすくなり、可読性や保守性の向上も見込めるようになります。

i-date commented on app/views/items/edit.html.erb:
```diff
@@ -0,0 +1,161 @@
+    <h2 class="items-sell-title">商品の情報を入力</h2>
+    <%= form_with model: @item, local: true do |f| %>
+
+    <%# インスタンスを渡して、エラー発生時にメッセージが表示されるようにしましょう。%>
+    <%= render 'shared/error_messages', model: @item %>
+    <%# //インスタンスを渡して、エラー発生時にメッセージが表示されるようにしましょう。%>
```
12、14行目の下記の実装案内コメントですが、機能を実装いただいたため、こちらは不要となりますので削除しましょう。
不要になったコメントを削除することで、可読性や保守性を向上させることができます。
```
<%# インスタンスを渡して、エラー発生時にメッセージが表示されるようにしましょう。%>
<%# //インスタンスを渡して、エラー発生時にメッセージが表示されるようにしましょう。%>
```
メモ代わりに残していただいている場合は修正は不要です。

kokoro-kusakabe replied:
変更箇所の修正が完了しましたので再度コードレビューを
お願いします

#出品者Aさんが商品First,商品Secondを出品し商品Secondだけ購入した状態
を用意しました。

Aさんが商品firstを編集画面に遷移し元々合った情報が入力された状態で表示されているか
  と戻るで遷移する前の画面になっているかの確認
　https://gyazo.com/0e0dead81d03dcf2c3b8ab49fe98b8bf
Aさんが商品firstを編集画面に遷移しimage以外空にした時のエラーメッセージの表示
　https://gyazo.com/27fe1616fc73965f10db4fd693a3e85f
Aさんが商品firstを編集した内容がしっかりと反映されているかの確認
　https://gyazo.com/5bd4f4be7a1b06d15e44231ca6b43367
Aさんが商品firstを編集しようとした時
　https://gyazo.com/24e4cdae988c330ef54bd2892bbf2605
Aさんが商品Secondを編集しようとした時
　https://gyazo.com/95413c7e68b61bf510944a23d6783d32
Bさんが商品firstを編集しようとした時
　https://gyazo.com/5b428942d69aa76925ec52cba205bfdf
Bさんが商品Secondを編集しようとした時
　https://gyazo.com/aa9ea2497bc90e65240935b0f5d4efef
ログアウトしてるユーザーが商品Firstを編集しようとした時
　https://gyazo.com/e214cd1513935b837d1996dbaa03b12c
ログアウトしてるユーザーが商品Secondを編集しようとした時
　https://gyazo.com/c138d75191721d34cdfb3ce65b079194

i-date reviewed (commented):
レビュー依頼とキャプチャ動画ありがとうございます！
コメントいたしましたので、ご確認をお願いいたします。

kokoro-kusakabe committed:
before_actionで同じ記述をまとめる

edit,showの部分を
@item = Item.find(params[:id])
def set_itemでまとめた

他にも同じ部分がある、
update,move_to_indexを編集した

kokoro-kusakabe committed:
コメントアウト部分の削除

kokoro-kusakabe committed:
同じ画面遷移するところを_pathにまとめました

i-date commented on app/controllers/items_controller.rb:
```diff
@@ -20,13 +22,41 @@ def create
+  end
+
+  def update
+    set_item
```
updateアクションも、3行目の`before_action :set_item, only: [:edit, :show]`に追加しましょう。
そうすることで、updateアクション内のset_itemを削除することができます。

kokoro-kusakabe replied:
再度修正を完了したのでコードレビューをお願いします。

i-date reviewed (commented):
修正ありがとうございます！
追加補足コメントいたしましたので、ご確認をお願いいたします。

kokoro-kusakabe committed:
update,move_to_indexをbefore_actionに追加

:move_to_indexも追加できるか試したら正常に機能したので
一応追加をしました。

i-date reviewed (approved):
修正ありがとうございます！
お疲れさまでした、LGTMとさせていただきます。
以降の実装も引き続きこの調子で進めていきましょう。応援しております！
